### Part 1: Scrape Toronto PostalCodes, Boroughs and Neighborhoods

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

Use requests to load website into `data` variable

In [2]:
url='https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969'
data = requests.get(url).text

Parse the html data using `beautiful_soup`.

In [122]:
soup = BeautifulSoup(data, "html5lib")

Find table that has "wikitable sortable" in class.

In [4]:
my_table = soup.find('table',{'class':'wikitable sortable'})

Scrape table and put into dataframe.

In [5]:
toronto_data = pd.DataFrame(columns=["PostalCode", "Borough", "Neighborhood"])

for row in my_table.find("tbody").find_all("tr"):
    col = row.find_all("td")
    if (col != []):
        PostalCode =col[0].text.strip()
        Borough =col[1].text.strip()
        Neighborhood =col[2].text.strip()
        toronto_data = toronto_data.append({"PostalCode":PostalCode, "Borough":Borough, "Neighborhood":Neighborhood}, ignore_index=True)

toronto_data
    

PostalCode           Borough  \
0          M1A      Not assigned   
1          M2A      Not assigned   
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
..         ...               ...   
175        M5Z      Not assigned   
176        M6Z      Not assigned   
177        M7Z      Not assigned   
178        M8Z         Etobicoke   
179        M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Clean Borough table and reset index

In [6]:
toronto_data['Borough']=toronto_data['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

toronto_data = toronto_data[toronto_data.Borough != 'Not assigned']

toronto_data.reset_index(drop=True, inplace=True)
toronto_data.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [7]:
toronto_data.shape

(103, 3)

The shape of the final table is 103 rows by 3 columns.

### Part 2: Add GeoSpatial Data

In [124]:
#!pip install geocoder  #install geocoder if necessary by removing #
import geocoder # import geocoder

User geocoder to return a list of the coordinates for each postalcode.

In [9]:
# initialize your variable to None
lat_lng_coords = None
gpslist =[]
i=0
postlist=toronto_data.PostalCode

for i in postlist:
    #loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(i))
        lat_lng_coords = g.latlng
    
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    print(i,latitude,longitude)
    gpslist.append([i,latitude,longitude])
    lat_lng_coords= None
    



M3A 43.75245000000007 -79.32990999999998
M4A 43.73057000000006 -79.31305999999995
M5A 43.65512000000007 -79.36263999999994
M6A 43.72327000000007 -79.45041999999995
M7A 43.66253000000006 -79.39187999999996
M9A 43.662630000000036 -79.52830999999998
M1B 43.811390000000074 -79.19661999999994
M3B 43.74923000000007 -79.36185999999998
M4B 43.70718000000005 -79.31191999999999
M5B 43.65739000000008 -79.37803999999994
M6B 43.70687000000004 -79.44811999999996
M9B 43.65034000000003 -79.55361999999997
M1C 43.78574000000003 -79.15874999999994
M3C 43.72168000000005 -79.34351999999996
M4C 43.68970000000007 -79.30681999999996
M5C 43.65215000000006 -79.37586999999996
M6C 43.69211000000007 -79.43035999999995
M9C 43.64857000000006 -79.57824999999997
M1E 43.765750000000025 -79.17469999999997
M4E 43.67709000000008 -79.29546999999997
M5E 43.64536000000004 -79.37305999999995
M6E 43.68784000000005 -79.45045999999996
M1G 43.76812000000007 -79.21760999999998
M4G 43.709020000000066 -79.36348999999996
M5G 43.65609

In [10]:
gpslist = pd.DataFrame(gpslist, columns=["PostalCode", "Latitude", "Longitude"])
gpslist

PostalCode  Latitude  Longitude
0          M3A  43.75245  -79.32991
1          M4A  43.73057  -79.31306
2          M5A  43.65512  -79.36264
3          M6A  43.72327  -79.45042
4          M7A  43.66253  -79.39188
..         ...       ...        ...
98         M8X  43.65319  -79.51113
99         M4Y  43.66659  -79.38133
100        M7Y  43.64869  -79.38544
101        M8Y  43.63278  -79.48945
102        M8Z  43.62513  -79.52681

[103 rows x 3 columns]

#### Final Combined Table of Neighborhoods and Geospatial Data

In [11]:
toronto_data=pd.merge(toronto_data, gpslist, on='PostalCode', how='left')
toronto_data

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  Latitude  Longitude  
0                                            Parkwoods  43.75245  -79.32991  
1                                     Victoria Village  43.73057  -79.31306  
2                            Regent Park, Harbourfront  43.65512  -79.36264  
3                     Lawrence Manor, Lawrence Heights  43.72327  -79.45042  
4          Queen's Park, Ontario Provincial Government  43.66253  -79.39188  
..                                                 ...       ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.65319  -79.51113  
99                                Church and Wellesley  43.66659  -79.38133  
100  Business reply mail Processing Centre, South C...  43.64869  -79.38544  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.63278  -79.48945  
102  Mimico NW, The Queensway West, South of Bloor,...  43.62513  -79.52681  

[103 rows x 5 columns]

### Part 3: Explore and Cluster the Neighborhoods in Toronto

In [12]:
import numpy as np # library to handle data in a vectorized manner

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!pip install geopy
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!pip install folium
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Function to return the nearby venues for each areacode.

In [13]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Foursquare Credentials and Version.  
**Note: Credentials removed for uploaded version.

In [1]:
CLIENT_ID = 'XXXX' # your Foursquare ID
CLIENT_SECRET = 'XXXX' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XXXX
CLIENT_SECRET:XXXX


Run the above function on each neighborhood group and create a new dataframe called `toronto_venues`.

In [19]:
toronto_data1 =toronto_data

toronto_venues = getNearbyVenues(names=toronto_data1['Neighborhood'],
                                   latitudes=toronto_data1['Latitude'],
                                   longitudes=toronto_data1['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [126]:
print("Shape of Table :",toronto_venues.shape)
toronto_venues.head()

Shape of Table : (2343, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods               43.75245               -79.32991   
1         Parkwoods               43.75245               -79.32991   
2         Parkwoods               43.75245               -79.32991   
3  Victoria Village               43.73057               -79.31306   
4  Victoria Village               43.73057               -79.31306   

                Venue  Venue Latitude  Venue Longitude     Venue Category  
0     Brookbanks Park       43.751976       -79.332140               Park  
1     Brookbanks Pool       43.751389       -79.332184               Pool  
2       Variety Store       43.751974       -79.333114  Food & Drink Shop  
3        Wigmore Park       43.731023       -79.310771               Park  
4  Memories of Africa       43.726602       -79.312427      Grocery Store

In [22]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                              15   
Alderwood, Long Branch                                                  4   
Bathurst Manor, Wilson Heights, Downsview North                         1   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      20   
Berczy Park                                                            65   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           85   
Business reply mail Processing Centre, South Ce...                     99   
CN Tower, King and Spadina, Railway Lands, Harb...                     75   
Caledonia-Fairbanks                                                     7   
Canada Post Gateway Processing Centre                                  99   
Cedarbrae                                                               2   
Central Bay Street                                                     60   
Christie                                                               11   
Church and Wellesley                                                   81   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         9   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             26   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              12   
Dorset Park, Wexford Heights, Scarborough Town ...                      4   
Downsview                                                              38   
Dufferin, Dovercourt Village                                           19   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      6   
Fairview, Henry Farm, Oriole                                           53   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         1   
Garden District, Ryerson                                              100   
Glencairn                                                              10   
Golden Mile, Clairlea, Oakridge                                        11   
Guildwood, Morningside, West Hill                                       4   
Harbourfront East, Union Station, Toronto Islands                      50   
High Park, The Junction South                                           3   
Hillcrest Village                                                       3   
Humber Summit                                                           1   
Humberlea, Emery                                                        4   
Humewood-Cedarvale                                                      5   
India Bazaar, The Beaches West                                         17   
Islington Avenue, Humber Valley Village                                 8   
Kennedy Park, Ionview, East Birchmount Park                             6   
Kensington Market, Chinatown, Grange Park                              45   
Kingsview Village, St. Phillips, Martin Grove G...                      3   
Lawrence Manor, Lawrence Heights                                       50   
Lawrence Park                                                           2   
Leaside                                                                32   
Little Portugal, Trinity                                               42   
Malvern, Rouge                                  

In [23]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 261 uniques categories.


Analyze Each Neighborhood.

In [24]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Zoo Exhibit  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   Auto Garage  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  Bakery  \
0            0          0           0                0           0       0   
1            0          0           0                0           0       0   
2            0          0           0                0           0       0   
3            0          0           0                0           0       0   
4            0          0           0                0           0       0   

   Bank  Bar  Baseball Field  Basketball Court  Basketball Stadium  Beer Bar  \
0     0    0               0                 0                   0         0   
1     0    0               0                 0                   0         0   
2     0    0               0                 0                   0         0   
3     0    0               0                 0                   0         0   
4     0    0               0                 0                   0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bike Trail  Bistro  \
0           0                   0          0           0       0   
1           0                   0          0           0       0   
2           0                   0          0           0       0   
3           0                   0          0           0       0   
4           0                   0          0           0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bridge  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0        0       0                0             0              0         0   
1        0       0                0             0              0         0   
2        0       0                0             0              0         0   
3        0       0                0             0              0         0   
4        0       0                0             0              0         0   

   Bus Station  Bus Stop  Business Service  Butcher  Café  Camera Store  \
0            0         0                 0        0     0             0   
1            0         0                 0        0     0             0   
2            0         0                 0        0     0             0   
3            0         0     

In [130]:
print("Dataframe Size: ",toronto_onehot.shape)

Dataframe Size:  (2343, 261)


Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [105]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Zoo Exhibit  \
0                                           Agincourt          0.0   
1                              Alderwood, Long Branch          0.0   
2     Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                     Bayview Village          0.0   
4                   Bedford Park, Lawrence Manor East          0.0   
5                                         Berczy Park          0.0   
6                         Birch Cliff, Cliffside West          0.0   
7        Brockton, Parkdale Village, Exhibition Place          0.0   
8   Business reply mail Processing Centre, South C...          0.0   
9   CN Tower, King and Spadina, Railway Lands, Har...          0.0   
10                                Caledonia-Fairbanks          0.0   
11              Canada Post Gateway Processing Centre          0.0   
12                                          Cedarbrae          0.0   
13                                 Central Bay Street          0.0   
14                                           Christie          0.0   
15                               Church and Wellesley          0.0   
16            Clarks Corners, Tam O'Shanter, Sullivan          0.0   
17    Cliffside, Cliffcrest, Scarborough Village West          0.0   
18                     Commerce Court, Victoria Hotel          0.0   
19                                         Davisville          0.0   
20                                   Davisville North          0.0   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn          0.0   
22                                          Don Mills          0.0   
23  Dorset Park, Wexford Heights, Scarborough Town...          0.0   
24                                          Downsview          0.0   
25                       Dufferin, Dovercourt Village          0.0   
26      East Toronto, Broadview North (Old East York)          0.0   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...          0.0   
28                       Fairview, Henry Farm, Oriole          0.0   
29             First Canadian Place, Underground city          0.0   
30    Forest Hill North & West, Forest Hill Road Park          0.0   
31                           Garden District, Ryerson          0.0   
32                                          Glencairn          0.0   
33                    Golden Mile, Clairlea, Oakridge          0.0   
34                  Guildwood, Morningside, West Hill          0.0   
35  Harbourfront East, Union Station, Toronto Islands          0.0   
36                      High Park, The Junction South          0.0   
37                                  Hillcrest Village          0.0   
38                                      Humber Summit          0.0   
39                                   Humberlea, Emery          0.0   
40                                 Humewood-Cedarvale          0.0   
41                     India Bazaar, The Beaches West          0.0   
42            Islington Avenue, Humber Valley Village          0.0   
43        Kennedy Park, Ionview, East Birchmount Park          0.0   
44          Kensington Market, Chinatown, Grange Park          0.0   
45  Kingsview Village, St. Phillips, Martin Grove ...          0.0   
46                   Lawrence Manor, Lawrence Heights          0.0   
47                                      Lawrence Park          0.0   
48                                            Leaside          0.0   
49                           Little Portugal, Trinity          0.0   
50                                     Malvern, Rouge          0.5   
51  Milliken, Agincourt North, Steeles East, L'Amo...          0.0   
52  Mimico NW, The Queensway West, South of Bloor,...          0.0   
53                        Moore Park, Summerhill East          0.0   
54       New Toronto, Mimico South, Humber Bay Shores          0.0   
55           North Park, Maple Leaf Park, Upwood Park          0.0   
56                 North Toronto West,  Lawrence Pa

In [132]:
print('Dataframe Size: ', toronto_grouped.shape)

Dataframe Size:  (97, 261)


Print each neighborhood along with the top 5 most common venues.

In [107]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                venue  freq
0  Chinese Restaurant  0.13
1      Discount Store  0.07
2    Sushi Restaurant  0.07
3              Bakery  0.07
4         Supermarket  0.07


----Alderwood, Long Branch----
                   venue  freq
0                    Pub  0.25
1      Convenience Store  0.25
2  Performing Arts Venue  0.25
3                    Gym  0.25
4      Mobile Phone Shop  0.00


----Bathurst Manor, Wilson Heights, Downsview North----
                      venue  freq
0               Men's Store   1.0
1               Zoo Exhibit   0.0
2            Massage Studio   0.0
3  Mediterranean Restaurant   0.0
4             Metro Station   0.0


----Bayview Village----
                        venue  freq
0                       Trail  0.50
1  Construction & Landscaping  0.25
2                        Park  0.25
3           Mobile Phone Shop  0.00
4               Movie Theater  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0         Coffee Sh

4                Modern European Restaurant  0.00


----Humber Summit----
                        venue  freq
0  Construction & Landscaping   1.0
1                 Zoo Exhibit   0.0
2     New American Restaurant   0.0
3                 Men's Store   0.0
4               Metro Station   0.0


----Humberlea, Emery----
               venue  freq
0        Coffee Shop  0.50
1               Park  0.25
2          Nightclub  0.25
3  Mobile Phone Shop  0.00
4      Movie Theater  0.00


----Humewood-Cedarvale----
           venue  freq
0   Hockey Arena   0.2
1           Park   0.2
2          Trail   0.2
3  Grocery Store   0.2
4          Field   0.2


----India Bazaar, The Beaches West----
                  venue  freq
0  Fast Food Restaurant  0.12
1            Restaurant  0.06
2                   Pub  0.06
3             Pet Store  0.06
4     Food & Drink Shop  0.06


----Islington Avenue, Humber Valley Village----
           venue  freq
0       Pharmacy  0.25
1  Grocery Store  0.12
2   Skating Ri

              venue  freq
0       Coffee Shop  0.07
1             Hotel  0.05
2              Café  0.04
3  Asian Restaurant  0.03
4  Sushi Restaurant  0.03


----Studio District----
                 venue  freq
0                Diner  0.07
1              Brewery  0.07
2               Bakery  0.07
3   Italian Restaurant  0.07
4  American Restaurant  0.04


----Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park----
                venue  freq
0  Light Rail Station  0.43
1         Coffee Shop  0.29
2         Supermarket  0.14
3        Liquor Store  0.14
4       Movie Theater  0.00


----The Annex, North Midtown, Yorkville----
            venue  freq
0  Sandwich Place  0.14
1            Café  0.10
2  History Museum  0.05
3       BBQ Joint  0.05
4        Pharmacy  0.05


----The Beaches----
               venue  freq
0                Pub   0.2
1   Asian Restaurant   0.2
2  Health Food Store   0.2
3              Trail   0.2
4      Movie Theater   0.0


----The Danforth West, R

#### Put into a pandas dataframe

Sort the venues in descending order.

In [108]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood.

In [109]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  1st Most Common Venue  \
0                                        Agincourt     Chinese Restaurant   
1                           Alderwood, Long Branch  Performing Arts Venue   
2  Bathurst Manor, Wilson Heights, Downsview North            Men's Store   
3                                  Bayview Village                  Trail   
4                Bedford Park, Lawrence Manor East     Italian Restaurant   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Supermarket          Skating Rink         Grocery Store   
1                         Gym     Convenience Store                   Pub   
2                 Yoga Studio                 Field        Farmers Market   
3  Construction & Landscaping                  Park           Yoga Studio   
4                 Coffee Shop        Sandwich Place           Sports Club   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0         Shopping Mall  Vietnamese Restaurant       Bubble Tea Shop   
1           Yoga Studio      Electronics Store               Dog Run   
2                  Farm     Falafel Restaurant           Event Space   
3     Electronics Store                Dog Run            Donut Shop   
4            Restaurant              Juice Bar     Indian Restaurant   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0        Discount Store  Hong Kong Restaurant                       Bakery  
1            Donut Shop           Dry Cleaner          Dumpling Restaurant  
2  Ethiopian Restaurant           Escape Room            Electronics Store  
3           Dry Cleaner   Dumpling Restaurant  Eastern European Restaurant  
4                   Pub       Thai Restaurant                         Café

#### Cluster Neighborhoods
Run _k_-means to cluster the neighborhood into 5 clusters.

In [110]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 4, 2, 0, 0, 0, 0, 0, 0])

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [111]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data
# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.75245  -79.32991             1.0     Food & Drink Shop   
1  43.73057  -79.31306             1.0              Pharmacy   
2  43.65512  -79.36264             0.0           Coffee Shop   
3  43.72327  -79.45042             0.0        Clothing Store   
4  43.66253  -79.39188             0.0           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0                  Park                    Pool     Electronics Store   
1                  Park           Grocery Store     German Restaurant   
2        Breakfast Spot                     Pub           Event Space   
3         Women's Store  Furniture / Home Store             Bookstore   
4    Falafel Restaurant               Gastropub                  Park   

         5th Most Common Venue     6th Most Common Venue  \
0                      Dog Run                Donut Shop   
1  Eastern European Restaurant           Doctor's Office   
2                          Spa         Electronics Store   
3             Toy / Game Store                Food Court   
4           Italian Restaurant  Mediterranean Restaurant   

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0           Dry Cleaner   Dumpling Restaurant  Eastern European Restaurant   
1               Dog Run            Donut Shop                  Dry Cleaner   
2            Food Truck            Restaurant           Italian Restaurant   
3           Men's Store        Cosmetics Shop                   Restaurant   
4         Burrito Place   Fried Chicken Joint                         Bank   

  10th Most Common Venue  
0            Yoga Studio  
1    Dumpling Restaurant  
2                 Bakery  
3    Leather Goods Store  
4         Sandwich Place

Drop rows with NA and change cluster labels to type int.

In [119]:
print('Before: ',toronto_merged.shape)
toronto_merged =toronto_merged.dropna()
toronto_merged = toronto_merged.reset_index(drop=True)
print('After: ',toronto_merged.shape)
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype('int64')

Before:  (101, 16)
After:  (101, 16)


#### Visualize Clusters

In [120]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

The above map shows the 5 clusters in Toronto.  Cluster 0 seems to contain the most neighborhoods.